File: Models/Item.swift
//
//  Item.swift
//  Machac
//
//  Created by Jan on 18/10/2017.
//  Copyright © 2017 Schwarja. All rights reserved.
//

import Foundation
import RealmSwift
import Realm

class Item: Object {
    @objc dynamic var id: String = ""
    @objc dynamic var name: String = ""
    @objc dynamic var owner: Person?
    @objc dynamic var valueInCurrency: Double = 0
    @objc dynamic var currency: Currency?
    let ratios = LinkingObjects(fromType: Ratio.self, property: "item")
    
    var value: Double {
        let czkValue = valueInCurrency * (currency?.relationToCzk ?? 0)

        if AppSettings.shared.referenceCurrency.isDefault {
            return czkValue
        }
        else {
            return czkValue / AppSettings.shared.referenceCurrency.relationToCzk
        }
    }
    
    override static func primaryKey() -> String? {
        return "id"
    }
    
    init(name: String, owner: Person, valueInCurrency: Double, currency: Currency, id: String? = nil) {
        self.id = id ?? NSUUID().uuidString
        self.name = name
        self.owner = owner
        self.valueInCurrency = valueInCurrency
        self.currency = currency
        
        super.init()
    }
    
    required init() {
        self.id = NSUUID().uuidString
        
        super.init()
    }
    
    required init(realm: RLMRealm, schema: RLMObjectSchema) {
        super.init(realm: realm, schema: schema)
    }
    
    required init(value: Any, schema: RLMSchema) {
        super.init(value: value, schema: schema)
    }
}

extension Item: CascadeDelete {
    
    func objectsToDelete() -> [Object] {
        var result: [Object] = []
        
        let ratios = Array(self.ratios)
        result.append(contentsOf: ratios as [Object])
        
        result.append(self)
        
        return result
    }
}

File: Android/app/src/main/java/com/example/janschwarz1/myapplication/models/Item.java
package com.example.janschwarz1.myapplication.models;

import com.example.janschwarz1.myapplication.utils.AppSettings;

import java.util.UUID;

import io.realm.RealmObject;
import io.realm.RealmResults;
import io.realm.annotations.LinkingObjects;
import io.realm.annotations.PrimaryKey;
import io.realm.annotations.Required;

/**
 * Created by janschwarz1 on 27/10/2017.
 */

public class Item extends RealmObject implements TitleValueAdapterItem, ItemWithCascadeDelete {
    @PrimaryKey@Required
    private String id;
    @Required
    private String name;
    private Person owner;
    @Required
    private Double valueInCurrency;
    private Currency currency;
    @LinkingObjects("item")
    private final RealmResults<Ratio> ratios = null;

    public Item() {
        this.id = UUID.randomUUID().toString();
    }

    public Item(String name, Person owner, Double valueInCurrency, Currency currency) {
        this(UUID.randomUUID().toString(), name, owner, valueInCurrency, currency);
    }

    public Item(String id, String name, Person owner, Double valueInCurrency, Currency currency) {
        this.id = id;
        this.name = name;
        this.owner = owner;
        this.valueInCurrency = valueInCurrency;
        this.currency = currency;
    }

    public Double value() {
        Double czkValue = valueInCurrency * currency.getRelationToCzk();
        if (AppSettings.shared.getReferenceCurrency().isDefault()) {
            return czkValue;
        } else {
            return czkValue / AppSettings.shared.getReferenceCurrency().getRelationToCzk();
        }
    }

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public Person getOwner() {
        return owner;
    }

    public Double getValueInCurrency() {
        return valueInCurrency;
    }

    public Currency getCurrency() {
        return currency;
    }

    public RealmResults<Ratio> getRatios() {
        return ratios;
    }

    @Override
    public String getTitle() {
        return name;
    }

    @Override
    public String getValue() {
        return "";
    }

    @Override
    public void cascadeDelete() {
        for (int i = ratios.size()-1; i >=0 ; i--) {
            ratios.get(i).cascadeDelete();
        }

        this.deleteFromRealm();
    }
}
